### PR TITLE
[Fix-18274][Registry] Handle expired JDBC heartbeat sessions

### LINKED ISSUE #18274: [Bug] [jdbc-registry] After database out of service for longer than session timeout time, client heartbeat will never be updated
### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

After database out of service for longer than session timeout time, client heartbeat will never be updated.

### What you expected to happen

If the database recover after session timeout time, client heatbeat should be updated again.

### How to reproduce

1. Shutdown database server for a session timeout time. 
2. Start database server.
3. If the scheduled task `purgeInvalidJdbcRegistryMetadata` purge the client heart beat before the heart beat be updated. After the client heart beat being deleted, the heart beat will never be updated by that client ID.

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qiuyanjun888**
## Was this PR generated or assisted by AI?

YES. This pull request was assisted by Hermes Agent / OpenAI Codex for code changes, focused tests, review feedback analysis, and local verification. The scope and final submission were directed by the contributor.

## Purpose of the pull request

Closes #18274.

When the JDBC registry database is unavailable longer than the session timeout, another server can purge the stale heartbeat row. After recovery, the still-running client previously ignored the zero-row heartbeat update and incorrectly treated the refresh as successful.

This is an independent alternative related to #18275. It addresses the outstanding technical concerns discussed there by not recreating or upserting an expired heartbeat, which could revive a failed-over identity. Instead, a missing heartbeat enters the existing disconnect state machine so the owning service can terminate.

## Brief change log

- Treat a zero-row JDBC heartbeat update as an expired registry session.
- Stop heartbeat refresh work after the registry server reaches `DISCONNECTED`.
- Persist the current heartbeat timestamp and add focused state-machine regression tests.

## Verify this pull request

This change added tests and can be verified as follows:

- `./mvnw clean -pl dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc -am -DskipITs -Dtest=JdbcRegistryServerTest,JdbcRegistryDataChangeListenerAdapterTest -Dsurefire.failIfNoSpecifiedTests=false test`
- Result: 4 tests, 0 failures, 0 errors, and `BUILD SUCCESS` across the selected reactor.
- `./mvnw -pl dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc spotless:apply`

## Pull Request Notice

[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

This pull request contains no incompatible change.

### CODE REVIEW

**Comment by SbloodyS** [comment]
When `updateById(clone)` returns `false`, the new `RegistryException` is caught by the generic error handler. If the server is currently `STARTED`, the catch block only moves it to `SUSPENDED`; it does not immediately invoke `onDisConnected()`.

This can happen after a long JVM pause or heartbeat scheduler starvation:

1. Another registry server considers the session expired and removes its heartbeat, ephemeral data, and locks.
2. The original server resumes while its local state is still `STARTED`.
3. `updateById()` returns `false`.
4. The original server enters `SUSPENDED` and remains active until at least the next heartbeat cycle.

A missing heartbeat row is definitive evidence that the session has expired, not a transient database error. Continuing to run after that point can overlap with a server that has already taken over, leaving a split-brain window.

Please handle this condition separately and transition directly to `DISCONNECTED`, triggering the disconnection callback exactly once.

The new test sets the server state to `SUSPENDED` and `lastSuccessHeartbeat` to zero, so it bypasses the problematic `STARTED` branch. Please also add a test verifying that a zero-row heartbeat update while `STARTED` causes an immediate disconnection.

**Comment by qiuyanjun888** [comment]
> When `updateById(clone)` returns `false`, the new `RegistryException` is caught by the generic error handler. If the server is currently `STARTED`, the catch block only moves it to `SUSPENDED`; it does not immediately invoke `onDisConnected()`.
> 
> This can happen after a long JVM pause or heartbeat scheduler starvation:
> 
> 1. Another registry server considers the session expired and removes its heartbeat, ephemeral data, and locks.
> 2. The original server resumes while its local state is still `STARTED`.
> 3. `updateById()` returns `false`.
> 4. The original server enters `SUSPENDED` and remains active until at least the next heartbeat cycle.
> 
> A missing heartbeat row is definitive evidence that the session has expired, not a transient database error. Continuing to run after that point can overlap with a server that has already taken over, leaving a split-brain window.
> 
> Please handle this condition separately and transition directly to `DISCONNECTED`, triggering the disconnection callback exactly once.
> 
> The new test sets the server state to `SUSPENDED` and `lastSuccessHeartbeat` to zero, so it bypasses the problematic `STARTED` branch. Please also add a test verifying that a zero-row heartbeat update while `STARTED` causes an immediate disconnection.

Thanks for your suggestions! I have already implemented, can you please help review?

**Comment by SbloodyS** [comment]
Thanks for addressing the previous feedback. A zero-row heartbeat update now disconnects the server immediately, and the added test covers the `STARTED` path.

However, there is still a shutdown race:

1. `refreshClientsHeartbeat()` passes the initial state check while the server is `STARTED`.
2. `close()` concurrently changes the state to `STOPPED` and deletes the heartbeat rows.
3. The in-flight `updateById()` returns `false`.
4. This branch unconditionally overwrites `STOPPED` with `DISCONNECTED` and invokes `onDisConnected()` during a normal shutdown.

Previously, the exception handler inspected the current state, so `STOPPED` was left unchanged. Please make the transition to `DISCONNECTED` conditional/atomic so that `close()` cannot race with the heartbeat task and have its terminal state overwritten. A regression test coordinating an in-flight heartbeat update with `close()` would also be useful.

**Comment by qiuyanjun888** [comment]
Addressed the shutdown race described in https://github.com/apache/dolphinscheduler/pull/18416#issuecomment-4998227869 in commit [`8d9599e`](https://github.com/qiuyanjun888/dolphinscheduler/commit/8d9599e097e062e1409a3c31dc1eaea4efa955cf).

- The transition to `DISCONNECTED` is now conditional and synchronized with `close()`, so a completed `STOPPED` transition cannot be overwritten and the disconnection callback is not invoked when close wins.
- Added a latch-controlled regression test that holds an in-flight heartbeat update, completes `close()`, then returns a zero-row update result.

Validation:
- `JdbcRegistryServerTest`: 5 tests, 0 failures, 0 errors
- module `spotless:check`: `BUILD SUCCESS`

Could you please take another look?

**Review by SbloodyS — COMMENTED** [review]
There is still a shutdown race in the other state transitions.

`close()` and `transitionToDisconnected()` are synchronized, but the successful heartbeat and exception paths still update `jdbcRegistryServerState` directly:

- `SUSPENDED -> STARTED` after a successful refresh
- `STARTED -> SUSPENDED` after a refresh exception

For example:

1. The heartbeat thread observes `SUSPENDED`.
2. `close()` changes the state to `STOPPED`.
3. The heartbeat thread writes `STARTED` and invokes `onReconnected()`.

The exception path can similarly overwrite `STOPPED` with `SUSPENDED`. Since the field is also neither `volatile` nor consistently accessed under the same lock, state visibility is not guaranteed.

Please make all state transitions atomic and use one synchronization strategy for every read/write, such as synchronized transition methods or an `AtomicReference` with compare-and-set. Connection callbacks should only run when their corresponding transition succeeds.

Please also add regression coverage for `close()` racing with:

- A successful heartbeat while the server is `SUSPENDED`.
- A heartbeat exception while the server is `STARTED`.

In both cases, `STOPPED` must remain the final state and no reconnect/disconnect callback should be triggered after shutdown.

**Comment by qiuyanjun888** [comment]
> Please make all state transitions atomic and use one synchronization strategy for every read/write. Connection callbacks should only run when their corresponding transition succeeds.

Thanks for pointing this out. Fixed in commit `79cc4a0cde`.

The server state now uses one synchronization strategy consistently:

- `start()`, `getServerState()`, the `close()` state update, and all heartbeat state transitions synchronize on the same server instance.
- Successful heartbeats use a guarded `SUSPENDED -> STARTED` transition.
- Heartbeat exceptions use a guarded `STARTED -> SUSPENDED` transition.
- Expired heartbeat records use a guarded `STARTED/SUSPENDED -> DISCONNECTED` transition.
- Reconnect and disconnect callbacks run inside their corresponding synchronized transition, only after the expected source state is confirmed. If `close()` wins and sets `STOPPED`, later heartbeat results cannot overwrite it or emit reconnect/disconnect callbacks.

I also added the two requested race regressions:

- `close()` racing with a successful heartbeat while `SUSPENDED`.
- `close()` racing with a heartbeat exception while `STARTED`.

Both assert that `STOPPED` remains final and that no reconnect/disconnect callback fires after shutdown.

Verification:

`./mvnw clean -pl dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc -am -DskipITs -Dtest=JdbcRegistryServerTest -Dsurefire.failIfNoSpecifiedTests=false test`

Result: 7 tests, 0 failures, 0 errors; reactor build succeeded.

**Review by SbloodyS — DISMISSED** [review]
LGTM

**Comment by SbloodyS** [comment]
PTAL @ruanwenjun

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 1]
```diff
@@ -256,7 +256,9 @@ public void releaseJdbcRegistryLock(Long clientId, String lockKey) {
 
     @Override
     public void close() {
-        jdbcRegistryServerState = JdbcRegistryServerState.STOPPED;
+        synchronized (this) {
+            jdbcRegistryServerState = JdbcRegistryServerState.STOPPED;
+        }
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:261]:
```suggestion
        synchronized (this) {
            if(jdbcRegistryServerState == JdbcRegistryServerState.STOPPED) {
                 log.warn("xx is already closed");
                 return
            }
            jdbcRegistryServerState = JdbcRegistryServerState.STOPPED;
        }
```

Comment (ruanwenjun) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:261]: Avoid duplicate `close`?

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 2]
```diff
@@ -341,27 +346,27 @@ private void refreshClientsHeartbeat() {
                 }
                 JdbcRegistryClientHeartbeatDTO clone = jdbcRegistryClientHeartbeatDTO.clone();
                 clone.setLastHeartbeatTime(now);
-                jdbcRegistryClientRepository.updateById(jdbcRegistryClientHeartbeatDTO);
+                if (!jdbcRegistryClientRepository.updateById(clone)) {
+                    log.error("The client heartbeat has expired: {}", jdbcRegistryClientHeartbeatDTO.getId());
+                    transitionToDisconnected();
+                    return;
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:352]:
```suggestion
                   throw xxException()
```

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 3]
```diff
@@ -370,6 +375,30 @@ private void refreshClientsHeartbeat() {
...
+    }
+
+    private synchronized void transitionToDisconnected() {
+        if (jdbcRegistryServerState != JdbcRegistryServerState.STARTED
+                && jdbcRegistryServerState != JdbcRegistryServerState.SUSPENDED) {
+            return;
+        }
+        jdbcRegistryServerState = JdbcRegistryServerState.DISCONNECTED;
+        doTriggerOnDisConnectedListener();
+    }
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:400]: Can we use `compareAndSwap` and `volatile` to reduce `synchronized`?

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:352]: Should consider session timeout.

Comment (qiuyanjun888) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java]: I checked this against the session lifecycle. updateById returning false means the heartbeat row no longer exists, and the normal purge path removes that row only after its stored session timeout has elapsed. The session has therefore already expired; waiting for another local timeout would keep a stale server active and reintroduce the split-brain window this PR fixes. I kept the immediate DISCONNECTED transition, with the STARTED zero-row regression test covering this case.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 1 above)
Comment (qiuyanjun888) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:265]: Addressed in fddfbe1a81. close() now atomically changes the server to STOPPED only once, and subsequent concurrent or repeated calls return without repeating scheduler shutdown or database cleanup. I also added a latch-controlled concurrent regression test that verifies the client purge runs only once.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 3 above)
Comment (qiuyanjun888) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java]: Addressed in fddfbe1a81. The state field is now volatile and heartbeat transitions use AtomicReferenceFieldUpdater.compareAndSet, so state reads and transition methods no longer require synchronized. Connection callbacks run only after the corresponding CAS succeeds. The small start/close lifecycle critical section is retained to prevent startup and shutdown from overlapping.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java]: Only a session timeout can cause the status to change to "disconnected".

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 4]
```diff
@@ -60,6 +61,10 @@
 @Slf4j
 public class JdbcRegistryServer implements IJdbcRegistryServer {
 
+    private static final AtomicReferenceFieldUpdater<JdbcRegistryServer, JdbcRegistryServerState> SERVER_STATE_UPDATER =
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java]: Use `AtomicReference`, don't write field name `jdbcRegistryServerState` here.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 5]
```diff
@@ -370,6 +384,33 @@ private void refreshClientsHeartbeat() {
...
+                    && currentState != JdbcRegistryServerState.SUSPENDED) {
+                return;
+            }
+            if (SERVER_STATE_UPDATER.compareAndSet(
+                    this, currentState, JdbcRegistryServerState.DISCONNECTED)) {
+                doTriggerOnDisConnectedListener();
+                return;
+            }
+        }
+    }
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:407]: Need to handle the status set failed case.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 2 above)
Comment (qiuyanjun888) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java]: Addressed in 71e8a14eb2. A failed heartbeat update now enters SUSPENDED first; the server transitions to DISCONNECTED only after the configured session timeout. The regression test covers the STARTED -> SUSPENDED -> DISCONNECTED sequence.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 4 above)
Comment (qiuyanjun888) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java]: Addressed in 71e8a14eb2. Server state is now stored in an AtomicReference, so reads and compare-and-set transitions no longer depend on AtomicReferenceFieldUpdater or a field-name string.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 5 above)
Comment (qiuyanjun888) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:407]: Addressed in 71e8a14eb2. CAS failures are now handled explicitly: the SUSPENDED transition records the current state, and the DISCONNECTED transition logs the failed attempt and retries after re-reading the state. The listener runs only after a successful state change.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 6]
```diff
@@ -256,7 +257,12 @@ public void releaseJdbcRegistryLock(Long clientId, String lockKey) {
 
     @Override
     public void close() {
-        jdbcRegistryServerState = JdbcRegistryServerState.STOPPED;
+        synchronized (this) {
+            if (serverState.getAndSet(JdbcRegistryServerState.STOPPED) == JdbcRegistryServerState.STOPPED) {
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:261]: Nit, use `compareAndSet`.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java (hunk 5 above)
Comment (ruanwenjun) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:407]: All `compareAndSet` failed should be handled.  Please don't reply to me with an AI response, you should at least know what the reviewer means.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 7]
```diff
@@ -341,27 +350,27 @@ private void refreshClientsHeartbeat() {
...
+                    throw new IllegalStateException(
+                            "The client heartbeat record no longer exists: " + jdbcRegistryClientHeartbeatDTO.getId());
+                }
                 jdbcRegistryClientHeartbeatDTO.setLastHeartbeatTime(clone.getLastHeartbeatTime());
             }
-            if (jdbcRegistryServerState == JdbcRegistryServerState.SUSPENDED) {
-                jdbcRegistryServerState = JdbcRegistryServerState.STARTED;
-                doTriggerReconnectedListener();
-            }
+            transitionToStarted();
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:360]: Don't add such method, looks so strange, this only transite from `SUSPENDED` to `STARTED.`

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java [hunk 8]
```diff
@@ -370,6 +379,33 @@ private void refreshClientsHeartbeat() {
...
+        if (serverState.compareAndSet(JdbcRegistryServerState.SUSPENDED, JdbcRegistryServerState.STARTED)) {
+            doTriggerReconnectedListener();
+        }
+    }
+
+    private void transitionToSuspended() {
+        if (!serverState.compareAndSet(JdbcRegistryServerState.STARTED, JdbcRegistryServerState.SUSPENDED)) {
+            log.debug("Failed to transition JdbcRegistryServer to SUSPENDED; current state is {}", serverState.get());
+        }
+    }
```
Comment (ruanwenjun) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-jdbc/src/main/java/org/apache/dolphinscheduler/plugin/registry/jdbc/server/JdbcRegistryServer.java:392]: Please remove this kind of method.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]
